docs/style-guide/typescript.md

```diff
@@ -0,0 +1,606 @@
+---
+layout: style-guide
+title: TypeScript Style Guide
+category: style-guide
+tags: [style-guide]
+order: 3
+---
+
+# TypeScript Style Guide
+
+Like many style guides, this TypeScript style guide exists for two primary
+reasons. The first is to provide guidelines that result in a consistent code
+style across all of the Enso codebases, while the second is to guide people
+towards a style that is expressive while still easy to read and understand.
+
+In general, it aims to create a set of 'zero-thought' rules in order to ease the
+programmer burden; there is usually only _one way_ to lay out code correctly.
+
+<!-- MarkdownTOC levels="2,3" autolink="true" -->
+
+- [TypeScript Style Guide](#typescript-style-guide)
+  - [Code Formatting](#code-formatting)
+    - [Line Width](#line-width)
+    - [Imports](#imports)
+    - [Sections](#sections)
+      - [An Example of Using Sections](#an-example-of-using-sections)
+    - [Vertical Spacing](#vertical-spacing)
+    - [Multi-Line Expressions](#multi-line-expressions)
+      - [Multi-Line Expression Examples](#multi-line-expression-examples)
+    - [Spacing](#spacing)
+      - [Spacing Examples as Function Definitions](#spacing-examples-as-function-definitions)
+  - [Naming](#naming)
+  - [Package Structure and Naming](#package-structure-and-naming)
+    - [Using Access Modifiers](#using-access-modifiers)
+  - [Build Tooling](#build-tooling)
+  - [Commenting](#commenting)
+    - [Documentation Comments](#documentation-comments)
+    - [Source Notes](#source-notes)
+    - [TODO Comments](#todo-comments)
+    - [Other Comment Usage](#other-comment-usage)
+  - [Program Design](#program-design)
+    - [Code Complexity](#code-complexity)
+    - [Safety](#safety)
+    - [Testing and Benchmarking](#testing-and-benchmarking)
+    - [Warnings, and Lints](#warnings-and-lints)
+      - [Warnings](#warnings)
+
+<!-- /MarkdownTOC -->
+
+## Code Formatting
+
+This section explains the rules for visually laying out your code. They provide
+a robust set of guidelines for creating a consistent visual to the code.
+
+Code style is _far_ more than just the visual formatting of the code, especially
+as formatting can often be automated. According to the documentation of rustfmt,
+"formatting code is a mostly mechanical task which takes both time and mental
+effort". While, in many cases, the programmer can be relieved of this burden
+through use of an automated formatter, it is sometimes the case that such a tool
+imposes _more_ cognitive load in programmers. With rustfmt, programmers tend to
+have to refactor long lines to use variables, and move code to specific modules
+or sections lest rustfmt produce code that is hard to read and write. Thus, it
+is very important to write code in such a way that we can be proud of its
+quality.
+
+### Line Width
+
+Each line in the source file must be of a maximum of 100 characters of text.
```

Comment:
`Each line in the source file must be of a maximum of 100 characters of text. (including comments)`